features/0474-protocol-role-request/index.md

```diff
@@ -0,0 +1,87 @@
+# Aries RFC 0474: Role Request 0.1
+- Authors: [Sam Curren](telegramsam@gmail.com)
+- Since: 2020-05-06
+- Status Note: initial proposal
+- Start Date: 2018-05-05 
+- Tags: feature, protocol
+
+## Summary
+
+This protocol requests the recipient to play a specified role in another protocol.
+
+## Motivation
+
+It is sometimes desirable to ask the other party to play a role within another protocol. Although some protocols already have a way to do this as an existing feature of the protocol, it is desirable have a general purpose way to communicate that intent.
+
+## Tutorial
+
+### Protocol Role Request 0.1
+
+URI: https://didcomm.org/protocol_role_request/0.1/<messageType>
+
+### Key Concepts
+
+This protocol is intentionally simple: further interactions either involve a message from the requested protocol or a problem report.
+
+### Roles
+
+**requester**: Sends the `please_play_role` message to the `requestee`
+
+**requestee**: Receives the `please_play_role` message from the `requestee`. Responds with a message from the specified protocol and role, or a `problem_report` indicating why that will not be done.
+
+### States
+
+**requested**: The _requester_ enters this state after sending a `please_play_role` message to the _requestee_.
+
+**accepted**: Both parties enter this state after the _requestee_ responds with a message from the indicated protocol and role.
+
+**rejected**: Both parties enter this state after the _requestee_ responds with a `problem_report` message.
+
+### Messages
+
+#### please_play_role
+
+```json
+{
+    "@type": "https://didcomm.org/protocol_role_request/0.1/please_play_role",
+    "@id": "<messageID>",
+    "protocol": "<ProtocolURI>",
+    "role": "<name of role within ProtocolURI>"
+}
+```
+
+
+
+**protocol**: The Protocol URI
+
+**role**: The role name specified in the specified protocol.
+
+Note that no other information is present in this message. If information is required to start the protocol in the role, it likely warrants a message within the protocol itself.
```

Comment:
This is exactly the realization I had with the proposed change to the Out-of-Band's `invitation` message - simply informing the other party what role the sender wishes them to perform is not enough. That role needs to be parametrized somehow. The sender including the request message is the most straightforward way to do this.

Think back to @dhh1128 's example from the last call: how do I ask you to connect and then bid on my Rembrandt? Asking you to play the role of *bidder* is not enough - I need to also instruct you to bid on the *Rembrandt*.

Functionally speaking, the out-of-band `request` message is enough for this purpose if we agree that parties may form  a permanent connection prior to the receiver sending the message.

Reply:
Two thoughts out of this:

- a `label` on the request could be sufficient.  Or we could reuse the `goal` and `goal-code` pair from OOB for that purpose.
- Once again we're back to OOB `request` "being sufficient" for everything, if we added the fields from the 'invitiation' message into it.  Or vice versa. We're doing development on OOB and it would be really nice to nail down the remaining issues with it.  My $0.02CDN is that we just have one OOB message and put everything and the kitchen sink in it :-). We keep moving in that direction.

Reply:
I agree on collapsing the two messages.

I don't think goal-code is sufficient to parametrize the role for the receiver. A parameter can be an identifier or a reference to a resource on the sender's side.

Reply:
@llorllale: I am conflicted. I agree with you that goal may not be enough. However, including the message has some conceptual problems. Explaining them will take a page or two of text. Let me do a writeup and then I'll come back here and link to what I create. Then we can discuss further.

Reply:
@llorllale (and @swcurran and @TelegramSam ): here is my answer about a more elegant way to do this: https://github.com/hyperledger/aries-rfcs/pull/482. Read RFC 0478 first, and 0482 second. This is only a draft, but I think it will spark an interesting conversation.